posts/alpine-stack-size.md

```diff
@@ -0,0 +1,172 @@
+
+* The stack size was being restricted within our CI-environment.
+* This was not a configuration issue backed into the Docker image, and changing Ruby configuration did not work (which it should).
+
+### Decreasing the stack size
+
+We then decided to reduce the stack size on our local machines (again with the `RUBY_THREAD_VM_STACK_SIZE` environment variable).
+
+What we discovered was that we needed to have stack sizes <500KB in order to reproduce the issue.
+
+This got us no closer to a fix, but did enable us to see the error locally and trace it to a recursive algorithm within RSpec.
+
+**What we knew now:**
+
+* The stack size was being restricted within our CI-environment.
+* This was not a configuration issue backed into the Docker image, and changing Ruby configuration did not work (which it should).
+* The configuration had an affect locally.
+
+### Running with Alpine locally
+
+Gaining some parity at this point made a lot of sense. While we didn't expect there to be a runtime difference within Ruby between local and Alpine, it was definitely worth ruling it out.
+
+Immediately we noticed the error locally.
+
+```
+SystemStackError: stack level too deep
+```
+
+**What we knew now:**
+
+* The stack size was being restricted within the Alpine docker image
+* The `RUBY_THREAD_VM_STACK_SIZE` environment variable had no affect within this docker image
+
+## What was wrong?
+
+Doing some google searching we ended up at musl-libc's FAQs, specifically the ["functional differences"](https://wiki.musl-libc.org/functional-differences-from-glibc.html#Thread-stack-size) i.e. incompatibilities with glibc. 
+
+What we learned is that musl provides a default stack size of 80k, which is a significant difference to the stacksize provided by glibc (generally 2MB-10MB).
+
+Moreover, it requires programs to explicitly ask for more (or less) stack via `pthread_attr_setstacksize`, which MRI Ruby does not appear to do.
+
+There is an open Ruby bug here: https://bugs.ruby-lang.org/issues/14387
+
+### What is RSpec doing?
+
+RSpec prints out a list of errors, and helpfully includes a snippet of source code from your test suite.
+
+```
+Failures:
+
+  1) 1 equals 2
+     Failure/Error: expect(1).to eq(2)
+
+       expected: 2
+            got: 1
+
+       (compared using ==)
+     # ./spec/test_spec.rb:12:in `block (11 levels) in <top (required)>'
+
+Finished in 0.01589 seconds (files took 0.07908 seconds to load)
+1 example, 1 failure
+
+Failed examples:
+
+rspec ./spec/test_spec.rb:11 # 1 equals 2
+```
+
+In the example above, the string `expect(1).to eq(2)` is extracted from your test suite source code.
+
+The way it does is that it parses your Ruby source code using [Ruby ripper](http://ruby-doc.org/stdlib-2.5.0/libdoc/ripper/rdoc/Ripper.html), and then searches for the relevant nodes in the Ruby language Abstract Syntax Tree (AST). 
+
+This requires a Depth-First-Search algorithm, which is more elegant when expressed as a recursive algorithm.
```

Comment:
Perhaps include something about the more context blocks, the more recursion occurred + therefore more stack was used.